### PR TITLE
[Proposal] export JS variables

### LINKED ISSUE #1531: Export CSS variable strings for JS consumption
I'm currently using [typestyle](https://github.com/typestyle/typestyle) in a couple of projects and am interested in having `variables.scss` available as an autogenerated typed JS importable equivalent.

Example:

```ts
import { Variables } from "@blueprintjs/core";

// some valid variables
// * Variables.PT_ELEVATION_SHADOW_0
// * Variables.PT_DIALOG_BOX_SHADOW
```

### PR DESCRIPTION
**Pull request by giladgray**
#### Changes proposed in this pull request:

- 🔥  `sass-variables` task is renamed to `variables`, and is now a task group (`variables-<package>` task for each package that defines the corresponding block in Gulpfile.js; currently just core)
- this task now generates `src/generated/variables.ts` from the bones of `variables.scss`
  - i got clever with the transformations, and special-cased a few things, but hey it works great!

#### Reviewers should focus on:

- thoughts how to make the logic more robust? lots of regular expressions...

### CODE REVIEW

File: gulp/sass.js [hunk 1]
```diff
@@ -21,15 +20,6 @@ module.exports = (blueprint, gulp, plugins) => {
         srcGlob: (project, excludePartials) => {
             return path.join(project.cwd, "src/**/", `${excludePartials ? "!(_)" : ""}*.scss`);
         },
-
-        // TODO: make this configurable from root
```
Comment (giladgray) [inline gulp/sass.js]: done! this is now totally configurable, and there will be a `variables-*` subtask for each package that defines a ` variables` block in Gulpfile.js

File: packages/core/src/common/variables.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+ * Licensed under the BSD-3 License as modified (the “License”); you may obtain a copy
+ * of the license at https://github.com/palantir/blueprint/blob/master/LICENSE
+ * and https://github.com/palantir/blueprint/blob/master/PATENTS
+ */
+
+export const { floor } = Math;
+
+/** JavaScript implementation of Sass' `rgba(color, alpha)` function that emits CSS `rgba(r, g, b, a)` string */
+export function rgba(color: string, alpha: number) {
+    const red = parseInt(color.slice(1, 3), 16);
```
Comment (adidahiya) [inline packages/core/src/common/variables.ts]: indexing into the string like this looks brittle

File: packages/core/src/common/index.ts [hunk 3]
```diff
@@ -12,15 +12,15 @@ export * from "./position";
...
-export const Keys = keys;
-export const Utils = utils;
+import * as Classes from "../common/classes";
+import * as Keys from "../common/keys";
+import * as Utils from "./utils";
 // NOTE: Errors is not exported in public API
 
+import * as Variables from "../generated/variables";
+
+export { Classes, Keys, Utils, Variables };
```
Comment (adidahiya) [inline packages/core/src/common/index.ts]: should be called `CSSVariables`, not simply `Variables`

**Comment by llorca** [comment]
Is there more thinking behind this proposal besides fixing #1531? A flat list seems subpar compared to the types of data structures we can leverage with JS

**Comment by llorca** [comment]
I propose we close this out and do some more fancy data structures for our visual design system in JS, somewhere around 2.0 -- some really interesting things to be done that don't necessarily require typestyle or CSS in JS support

**Comment by giladgray** [comment]
@llorca agreed!

**Pull request closed without merging**